store/sqlstore/channel_member_history_store.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package sqlstore
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/store"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+type SqlChannelMemberHistoryStore struct {
+	SqlStore
+}
+
+func NewSqlChannelMemberHistoryStore(sqlStore SqlStore) store.ChannelMemberHistoryStore {
+	s := &SqlChannelMemberHistoryStore{
+		SqlStore: sqlStore,
+	}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.ChannelMemberHistory{}, "ChannelMemberHistory").SetKeys(false, "ChannelId", "UserId", "JoinTime")
+		table.ColMap("ChannelId").SetMaxSize(26)
+		table.ColMap("UserId").SetMaxSize(26)
+		table.ColMap("JoinTime").SetNotNull(true)
+	}
+
+	return s
+}
+
+func (s SqlChannelMemberHistoryStore) LogJoinEvent(userId string, channelId string, joinTime int64) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		channelMemberHistory := &model.ChannelMemberHistory{
```

Comment:
Generally we have the caller construct the object.

Reply:
Why? Constructing the object in the caller violates the Law of Demeter (principle of least knowledge) - the caller doesn't care what type of object we're putting in the database, or even that the database exists. It just knows that it wants to log an event, and that a user id, channel id, and join time are required to do so.

Reply:
You make a good argument. Had to think about it a bit. I think there is an argument for both ways. For bigger objects like user/channel objects I think the answer is more clear. For this? I am fine with leaving it like this because I doubt the interface will change much.

Reply:
Users of this store already have to know what type of object is going into and being pulled out of the database since the getters return `model.ChannelMemberHistory` objects. Given that, I'm on the side of having the caller construct the object so we can have named parameters:

```go
// I'm certainly not gonna know which of these is correct without looking it up:
s.LogJoinEvent(channelId, userId, now)
s.LogJoinEvent(userId, channelId, now)

// But now everything is crystal clear!
s.LogJoinEvent(&model.ChannelMemberHistory{
    UserId:    userId,
    ChannelId: channelId,
    JoinTime:  now,
})
```